llvm/include/llvm/CodeGen/TargetLowering.h

```diff
@@ -2975,6 +2975,8 @@ class TargetLoweringBase {
     return isTruncateFree(Val.getValueType(), VT2);
   }
 
+  virtual bool shouldReduceRegisterPressure() const { return false; }
```

Comment:
Please do not add any vague, argument-free, undocumented target lowering hooks. We have far too many single purpose random hooks like this. If it has no arguments, it probably should not be a virtual. 

Every target would claim yes, this needs to be more specific why. From the context, this is probably already covered by existing APIs (probably isTruncateFree)?

Reply:
Implementing `IsTruncateFree` may, indeed be sufficient in this case, though it addresses only this specific issue. 

We want to be able to make similar decisions in other cases, and it's not specific to a particular instruction or type, it's a target-wide parameter. I'm not quite sure how we would parameterize that.

The register pressure issue in NVPTX is, unfortunately, both very real (it's one of the most common performance regression cliffs for the GPUs), and it's also not visible on IR or PTX level, as the compilation for the actual GPU code, and eventual register spilling, is done by ptxas, outside of LLVM. All we can do is try using less registers and limit their lifetime. I can't think of a good way to quantify it, other than the vague "use less registers, please....". I think NVPTX is somewhat unique among LLVM back-ends in that regard.

Reply:
Have you tried removing it? The isZExtFree check is usually to limit the fold on x86 (where only i32->i64 is typically free) - some of the time its actually superfluous.

Reply:
The API is still bad and single use. This is essentially the same thing as aggressivelyPreferBuildVectorSources, and isNarrowingProfitable

Reply:
> This is essentially the same thing as aggressivelyPreferBuildVectorSources, and isNarrowingProfitable

Thank you for pointing this out! I agree that `isNarrowingProfitable()` is the right way forward.